src/Tools/dotnet-monitor/CollectionRules/ActionListExecutor.cs

```diff
@@ -33,47 +35,100 @@ public ActionListExecutor(ILogger<CollectionRuleService> logger, ICollectionRule
 
-                try
-                {
-                    ICollectionRuleActionFactoryProxy factory;
+                    _logger.CollectionRuleActionStarted(context.Name, actionOption.Type);
 
-                    if (!_actionOperations.TryCreateFactory(actionOption.Type, out factory))
+                    try
+                    {
+                        ICollectionRuleActionFactoryProxy factory;
+
+                        if (!_actionOperations.TryCreateFactory(actionOption.Type, out factory))
+                        {
+                            throw new InvalidOperationException(Strings.ErrorMessage_CouldNotMapToAction);
+                        }
+
+                        ICollectionRuleAction action = factory.Create(context.EndpointInfo, actionOption.Settings);
+
+                        try
+                        {
+                            await action.StartAsync(cancellationToken);
+
+                            // Check if the action completion should be awaited synchronously (in respected to
+                            // starting the next action). If not, add a deferred entry so that it can be completed
+                            // add the end of the execution list.
+                            if (actionOption.WaitForCompletion.GetValueOrDefault(CollectionRuleActionOptionsDefaults.WaitForCompletion))
+                            {
+                                await action.WaitForCompletionAsync(cancellationToken);
+
+                                _logger.CollectionRuleActionCompleted(context.Name, actionOption.Type);
+                            }
+                            else
+                            {
+                                deferredCompletions.Add(new(action, actionOption, actionIndex));
+
+                                // Set action to null to skip disposal
+                                action = null;
+                            }
+                        }
+                        finally
+                        {
+                            await DisposableHelper.DisposeAsync(action);
+                        }
+                    }
+                    catch (Exception ex) when (ShouldHandleException(ex, context.Name, actionOption.Type))
                     {
-                        throw new InvalidOperationException(Strings.ErrorMessage_CouldNotMapToAction);
+                        throw new CollectionRuleActionExecutionException(ex, actionOption.Type, actionIndex);
                     }
 
-                    ICollectionRuleAction action = factory.Create(context.EndpointInfo, actionOption.Settings);
+                    ++actionIndex;
+                }
+
+                // Notify that all actions have started
+                startCallback?.Invoke();
 
+                // Wait for any actions whose completion has been deferred.
+                for (int i = 0; i < deferredCompletions.Count; i++)
```

Comment:
More-so a thought than asking for a change, but I was curious about why we don't use a while loop here (e.g. `while (deferredCompletions.Count > 0`). Then we wouldn't have to play the game with incrementing and decrementing `i`, and when we assign `deferredCompletion` we could always use `deferredCompletions[0]`. In my head at least that feels more intuitive, but this also works.

Reply:
nit: Alternatively consider a queue